src/core/app.lua

```diff
@@ -229,27 +216,79 @@ function apply_config_actions (actions, conf)
       if not new_app_table[ta] then error("no such app: " .. ta) end
       -- Create or reuse a link and assign/update receiving app index
       local link = link_table[linkspec] or link.new(linkspec)
-      link.receiving_app = app_name_to_index[ta]
       -- Add link to apps
       new_app_table[fa].output[fl] = link
       table.insert(new_app_table[fa].output, link)
       new_app_table[ta].input[tl] = link
       table.insert(new_app_table[ta].input, link)
       -- Remember link
       new_link_table[linkspec] = link
-      table.insert(new_link_array, link)
    end
    -- Free obsolete links.
    for linkspec, r in pairs(link_table) do
       if not new_link_table[linkspec] then link.free(r, linkspec) end
    end
    -- Commit changes.
    app_table, link_table = new_app_table, new_link_table
-   app_array, link_array = new_app_array, new_link_array
    -- Trigger link event for each app.
-   for _, app in ipairs(app_array) do
+   for name, app in pairs(app_table) do
       if app.link then app:link() end
    end
+   compute_breathe_order ()
+end
+
+-- Sort the NODES topologically according to SUCCESSORS via
+-- reverse-post-order numbering.  The sort starts with ENTRIES.  This
+-- implementation is recursive; we should change it to be iterative
+-- instead.
+function tsort (nodes, entries, successors)
+   local visited = {}
+   local post_order = {}
+   local function visit(node)
+      visited[node] = true
+      for _,succ in ipairs(successors[node]) do
+         if not visited[succ] then visit(succ) end
+      end
+      table.insert(post_order, node)
+   end
+   for _,node in ipairs(entries) do
+      if not visited[node] then visit(node) end
+   end
+   for name,node in pairs(nodes) do
+      if not visited[node] then visit(node) end
+   end
+   local ret = {}
+   while #post_order > 0 do
+      table.insert(ret, table.remove(post_order))
+   end
+   return ret
+end
+
+breathe_pull_order = {}
```

Comment:
Well... in my patch it isn't.  This is more of a set; you want pulls to come first.  Probably the ideal is to run all pulls needed by each chain; if there are parallel disjoint graphs you should run pull-1 then all pushes in that chain, then pull-2 then all pushes in that chain, and so on.  Right now the patch does the simple thing though.  It could be that pull-1 and pull-2 are both consumed by push-N though, so it's a reasonable first step.